aesd-char-driver/main.c

```diff
@@ -0,0 +1,307 @@
+    if (buf == NULL) { //error condition
+	    //TODO: assert?
+        return NULL;
+    }
+
+    //copy old  and new contents into newly allocated memory and free the old memory
+    memcpy(buf, partial_data_buf, partial_mem_size);
+    memcpy((buf+partial_mem_size), new_data_ptr, (new_size-partial_mem_size));
+
+    kfree(partial_data_buf);
+
+    return buf;
+}
+
+/** check if the given string is '\n' terminated.
+ *
+ *  @return 1 is not terminated
+ *          0 if terminated 
+ **/
+bool is_data_partial(char *buffptr, size_t count)
+{
+    int n = 0;
+    int len = 0;
+    bool partial = 1;
+
+    for (n = 0; n < count; n++) {
+	len++;
+        if (buffptr[n] == '\n') { //if '\n' is found, set partial flag to 0
+	    partial = 0;
+        }
+    }
+
+    if ((partial == 0) && (len != count)) {
+	PDEBUG("is_data_partial:partial pkt length error\r\n"); //TODO: improvise the error
+    }
+
+    return partial;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    struct aesd_dev *dev = filp->private_data;
+    struct aesd_buffer_entry wr_entry;
+    char  *rm_buffptr = NULL; //removed entry
+    char   *buffptr;
+    bool   partial_data = 1;
+    int    total_buf_size = 0;
+
+    ssize_t retval = -ENOMEM;
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+
+    //check the write data size(is this check needed?)
+    if (count > 1024) {
+        PDEBUG(" Size to be written is too large\r\n");	
+	//TODO: return error ?
+    }
+
+    buffptr = kmalloc(count, GFP_KERNEL);
+    if (buffptr == NULL) {
+        PDEBUG(" aesd_write: Failed to allocate memory for data\r\n");
+        return -ENOMEM;
+    }
+
+    //copy from user
+    if (__copy_from_user(buffptr, buf, count) != 0) {
+	PDEBUG(" aesd_write: failed to copy user data\r\n");
+	//TODO: return value?
+    }
+    
+    //check if data is \n terminated
+    partial_data = is_data_partial(buffptr, count);
+    if (partial_data == 0) {
+
+	// if there is data in partial buffer, handle that data
+	if (dev->part_mem_ptr != NULL) {
+            buffptr = handle_partial_data(dev->part_mem_ptr, 
+			    dev->part_buf_size, buffptr, (dev->part_buf_size+count));
+	    total_buf_size = dev->part_buf_size+count;
+	    //reset partial buffer info
+	    dev->part_mem_ptr = NULL;
+	    dev->part_buf_size = 0;
+	}
+
+	PDEBUG(" aesd_write: FINAL BUFF contains %s and count is %d\r\n", buffptr, count);
+	
+	wr_entry.buffptr = buffptr;
+	wr_entry.size = (total_buf_size != 0) ? total_buf_size:count;//handle partial data if any
+	
+	//take lock before writting the data into circular buffer
+	if (mutex_lock_interruptible(&dev->buf_lock)) {
+	    kfree(buffptr);
+            return 0;//-ERESTARTSYS;
+	}
+
+	rm_buffptr = aesd_circular_buffer_add_entry(&dev->buffer, &wr_entry);
+	mutex_unlock(&dev->buf_lock);
+
+        if (rm_buffptr != NULL) { //overwritten entry, free it
+	   kfree(rm_buffptr);
```

Comment:
@maheshchnt think about how the code here could interact with the code at https://github.com/cu-ecen-aeld/assignments-3-and-later-maheshchnt/pull/6/files#diff-e6d206de490054ca882e24fc06e3752cfc8240c67c2527207d6c56624f478901R80 when not mutex locked.